### ISSUE TITLE
Question about DB locking

### ISSUE THREAD

**Issue by beegmon** [ConsenSys/web3signer#558]
Hello,

I run a few fairly large validator sets for both eth and gnosis right now and am looking at Web3Signer as a way to help ease some of the scaling issue that come with localized slashing DB. When I say large validator sets I am talking about keyset sizes north of the 10K range for eth and gno.

Ideally, I want to move all my keys to Web3Signer running in an HA cluster with shared backend DB cluster that is also HA. This removes a lot of complexity out of my validator setup as they don't have to keep local slashing DBs on a per validator basis.

This brings me to my question. A quick glance of the code, and I do mean quick, tells me there are two ways the DB is locked. Per my understanding when a new key is added/removed/validator activated the entire DB is locked (probably wrong about this but that is my assumption).

Under normal block proposal/attestation process, again per my  glance, the code seems to only lock the rows of the specific validatorID involved, not the entire DB. Thus, multiple keys can be attesting/proposing at once and there is no lock contention across the the DB but instead on a per key basis.

To put it more simply, the lock in the DB for block proposal/attestations is per key, not the entire DB. Where as the lock when adding/removing/activating a validator is across the entire DB.

Finally, has the team done any load testing to see how Web3Signer scales up in regards to key sizes?

Thanks and looking forward to being corrected in regard to my assumptions about the DB locking in the signer.

**Comment by jframe** [comment]
Hi @beegmon,

There are two distinct cases for locking.

In the normal block and attestation signing we have database lock that is specific the validator id and the type of data being signed i.e. block or attestation. So multiple keys can be attesting/proposing at once so long as they are not for the same validator id and operation.

For the other case for importing or deleting a key using the key manager API we also use database locking. In this case it's not as fine grained in that we lock the database for both attestations and blocks for a particular validator id. This is not database wide lock it still specific the validator id.

We haven't done any load testing against large key sizes. This is something that we are working on though and currently in the process of setting a test with 10,000 keys.

**Comment by beegmon** [comment]
Thanks for the info and very much appreciated.
